python-checks/src/main/java/org/sonar/python/checks/ItemOperationsTypeCheck.java

```diff
@@ -39,37 +45,61 @@ public class ItemOperationsTypeCheck extends ItemOperationsType {
 
   @Override
   public boolean isValidSubscription(Expression subscriptionObject, String requiredMethod, @Nullable String classRequiredMethod,
-                                     Map<LocationInFile, String> secondaries) {
+    Map<LocationInFile, String> secondaries) {
 
     if (subscriptionObject.is(Tree.Kind.GENERATOR_EXPR)) {
       return false;
     }
-    if (subscriptionObject.is(Tree.Kind.CALL_EXPR)) {
-      Symbol subscriptionCalleeSymbol = ((CallExpression) subscriptionObject).calleeSymbol();
-      if (subscriptionCalleeSymbol != null && subscriptionCalleeSymbol.is(FUNCTION) && ((FunctionSymbol) subscriptionCalleeSymbol).isAsynchronous()) {
-        FunctionSymbol functionSymbol = (FunctionSymbol) subscriptionCalleeSymbol;
-        secondaries.put(functionSymbol.definitionLocation(), String.format(SECONDARY_MESSAGE, functionSymbol.name()));
-        return false;
-      }
+    if (isInvalidSubscriptionCallExpr(subscriptionObject, secondaries)) {
+      return false;
     }
-    if (subscriptionObject instanceof HasSymbol hasSymbol) {
-      Symbol symbol = hasSymbol.symbol();
-      if (symbol == null || isTypingOrCollectionsSymbol(symbol)) {
-        return true;
-      }
-      if (symbol.is(FUNCTION, CLASS)) {
-        secondaries.put(symbol.is(FUNCTION) ?
-          ((FunctionSymbol) symbol).definitionLocation() : ((ClassSymbol) symbol).definitionLocation(), String.format(SECONDARY_MESSAGE, symbol.name()));
-        return canHaveMethod(symbol, requiredMethod, classRequiredMethod);
-      }
+
+    var symbol = TreeUtils.getSymbolFromTree(subscriptionObject)
+      .filter(s -> s.is(FUNCTION, CLASS))
```

Comment:
as discussed - the previous logic was to return `true` even if a symbol is not a function or a class but has an `fqn` starting from `typing` or `collection`, need to revisit this condition